dhall-docs/src/Dhall/Docs/CodeRenderer.hs

```diff
@@ -0,0 +1,183 @@
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+-- >>> import Dhall.Core (Directory (..))
+
+-- | Given a Dhall expression return all imports with their location on the file.
+--   Contents are already sorted by 'Src', which allows 'renderAsHtml' to traverse
+--   the return of this function to inject anchors
+getImports :: Expr Src Import -> [(Src, Import)]
+getImports (Note src (Embed a)) = [(src, a)]
+getImports expr = concatMap getImports $ Lens.toListOf Core.subExpressions expr
+
+fileAsText :: File -> Text
+fileAsText File{..} = foldr (\d acc -> acc <> "/" <> d) "" (Core.components directory)
+    <> "/" <> file
+
+{-| Given an 'Import', render the contents in an HTML element that will allow
+    users to jump to another file or domain. The 'Text' argument is the contents
+    inside the anchor tag
+
+    Example:
+
+    >>> :set -Wno-missing-fields
+    >>> let file = File { directory = Directory [], file = ""}
+    >>> let url = URL { scheme = HTTPS, authority = "google.com", query = Nothing, path = file}
+    >>> let import_ = Import {importHashed = ImportHashed { importType = Remote url }}
+    >>> renderImport import_ "link for google"
+    <a href="https://google.com/" target="_blank">link for google</a>
+-}
+renderImport :: Import -> Text -> Html ()
+renderImport (Import {importHashed = ImportHashed { importType }}) =
+    case importType of
+        Remote URL {..} -> a_ [href_ href, target_ "_blank"] . toHtml
+          where
+            scheme_ = case scheme of
+                HTTP -> "http"
+                HTTPS -> "https"
+
+            path_ = fileAsText path
+
+            query_ = case query of
+                Nothing -> ""
+                Just d -> "?" <> d
+
+            -- we don't include the headers here since we treat links to open a file
+            -- in another tab
+            href = scheme_ <> "://" <> authority <> path_ <> query_
+
+        Local Here file -> a_ [href_ href] . toHtml
+          where
+            href = "." <> fileAsText file <> ".html"
+
+        Local Parent file -> a_ [href_ href] . toHtml
+          where
+            href = ".." <> fileAsText file <> ".html"
+
+        _ -> toHtml
+
+-- | Given a Text and the parsed `Expr Src Import` from it, this will render the
+--   the source code on HTML with jump-to-definition on URL imports. Use this
+--   to render the source code with the same structure (whitespaces, comments,
+--   language elements) as the source file
+exprSrcToHtml :: Text -> Expr Src Import -> Html ()
+exprSrcToHtml contents expr = pre_ $ go (1, 1) (Data.Text.lines contents) imports
+  where
+    imports = getImports expr
+
+    -- we keep the current line, column and consumed text as part of function argument
+    go :: (Int, Int) -> [Text] -> [(Src, Import)] -> Html ()
+    go _ textLines [] = mapM_ (\t -> toHtml t >> br_ []) textLines
+    go (currLine, currCol) currentLines ((Src {..}, import_) : rest) = do
```

Comment:
This still feels a bit convoluted, but I think I get the general idea.

Maybe it would be worth splitting out a pure function that takes care of the alignment of imports. Roughly like

```haskell
alignAnnotations
    :: Text
    -> [(Int, ann)] -- ^ (Column number, annotation), column numbers must be ascending
    -> [(Text, [ann])] -- ^ Lines of the original Text, with annotations for each column
```

Then you can just fold over the resulting annotated lines and handle them in a uniform fashion.

Reply:
Oh, in `alignAnnotations`, what I called the column numbers actually are the _line_ numbers. I had mixed up the terms!

Reply:
Sorry @sjakobi, but I don't get your idea :confused:

Reply:
what I do in a high-level is _traversing_ the text, using the first argument of `go` as the _cursor_. The `[Text]` argument is used as the remainder of the text consume, the first line is not be the _whole_ line if the _cursor_ column is different than `1`.

We consume the prefix of the import position an render as plain-text, render the import using a link if makes sense, an finally recur with the remainder of the text.

Reply:
(Probably this idea is what you suggested at first glance, sorry for that!)

We could take the source text and _partitionate_ it using a custom data type:

```haskell
data SourceCodeAnn = Normal Text | Link Text Import

-- if we concatenate the `SourceCodeAnn`'s `Text` we should get the input `Text`
partitionate :: Text -> [SourceCodeAnn]

render :: [SourceCodeAnn] -> Html ()
```

Reply:
Yes, that's what I meant. `SourceCodeAnn` is a homomorphism of `(Text, Maybe Import)`. I had suggested multiple imports / annotations because each line can contain multiple imports. When imports can also span multiple lines, things probably get a bit more complicated. No idea whether it's worthwhile.

Reply:
I can try this in a follow-up PR. Remember that we also want to support jump-to-definition on other language elements, so probably this code will suffer changes.